iOS_SDK/OneSignalSDK/Source/OneSignalLocation.m

```diff
@@ -186,15 +187,17 @@ + (void) internalGetLocation:(bool)prompt {
         //Location Always requires: Location Background Mode + NSLocationAlwaysUsageDescription
         NSArray* backgroundModes = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIBackgroundModes"];
         NSString* alwaysDescription = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"NSLocationAlwaysUsageDescription"] ?: [[NSBundle mainBundle] objectForInfoDictionaryKey:@"NSLocationAlwaysAndWhenInUseUsageDescription"];
-        if(backgroundModes && [backgroundModes containsObject:@"location"] && alwaysDescription) {
+        // use background location updates if always permission granted or prompt allowed
+        if(backgroundModes && [backgroundModes containsObject:@"location"] && alwaysDescription && (permissionStatus == 3 || prompt)) {
             [locationManager performSelector:@selector(requestAlwaysAuthorization)];
             if (deviceOSVersion >= 9.0) {
                 [locationManager setValue:@YES forKey:@"allowsBackgroundLocationUpdates"];
             }
         }
         
-        else if([[NSBundle mainBundle] objectForInfoDictionaryKey:@"NSLocationWhenInUseUsageDescription"])
-            [locationManager performSelector:@selector(requestWhenInUseAuthorization)];
+        else if([[NSBundle mainBundle] objectForInfoDictionaryKey:@"NSLocationWhenInUseUsageDescription"]) {
+            if (permissionStatus == 0) [locationManager performSelector:@selector(requestWhenInUseAuthorization)];
```

Comment:
We may also want to check for `kCLAuthorizationStatusRestricted` here.

Reply:
Why is that?
> `kCLAuthorizationStatusRestricted`: This app is not authorized to use location services.
The user cannot change this app’s status, possibly due to active restrictions such as parental controls being in place.

The theme I understood from the code is that if the app doesn't have permission, just continue on anyway and the calls to start location tracking will simply do nothing.

Reply:
You are right, we only need to check for `0` here. Was thinking the inverse at first.